File: restore/server/src/main/java/bgu/spl/net/impl/Frames/MessageFrame.java
package bgu.spl.net.impl.Frames;

import java.util.Map;

import bgu.spl.net.impl.User.User;
import bgu.spl.net.impl.gamesUpdate.Event;
import bgu.spl.net.impl.gamesUpdate.GamesFeed;
import bgu.spl.net.impl.stomp.StompCommand;
import bgu.spl.net.impl.stomp.StompFrame;

public class MessageFrame extends StompFrame{
   
    String destination;
    String subscription;
    int msgId;
    Event event; // maybe a list ?

    public MessageFrame(Map<String,String> headers) {
        super(StompCommand.MESSAGE, headers);
    }

    @Override
    public StompFrame create(String[] bodyFields){
        if(!stompsHeaders.containsKey("message-id")){
            missingHeaders.add("message-id");
            errorAccourd = true;
        }
        if(!stompsHeaders.containsKey("subscription")){
            missingHeaders.add("subscription");
            errorAccourd = true;
        }
        if(!stompsHeaders.containsKey("destination")){
            missingHeaders.add("destination");
            errorAccourd = true;
        }

        if(stompsHeaders.containsKey("receipt")){
            receiptId = stompsHeaders.get("receipt");
            errorHeaders.put("receipt", receiptId);
       }
       if(errorAccourd){
            return frameUtil.buildHeaderMissingErrorFrame(errorHeaders,missingHeaders);
       }

       msgId = Integer.parseInt(stompsHeaders.get("message-id"));
       subscription = stompsHeaders.get("subscription");
       destination = stompsHeaders.get("destination");

        // frame body
        FrameUtil frameUtil = new FrameUtil();
        Map<String,String> bodyheaders = frameUtil.getBodyContent(bodyFields);
        Map<String,String> props = frameUtil.getGameGeneralUpdates(bodyheaders.get("general game updates"));
        Map<String,String> aUpdates = frameUtil.getTeamUpdates(bodyheaders.get("team A updates"));
        Map<String,String> bUpdates = frameUtil.getTeamUpdates(bodyheaders.get("team B updates"));

        event = frameUtil.assignToFields(bodyheaders, props, aUpdates, bUpdates);
        body = event;
        return this;
    }


    @Override
    public void excute(GamesFeed gamesFeed, User user){
        gamesFeed.connections.send(destination, this);
    }
}

File: string to event bug/client/src/Frames/MessageFrame.cpp
#include "../include/StompFrame.h"
#include <iostream>
using std::cout;

//server send to client
MessageFrame::MessageFrame(): destination(""), subscription(-1), messageId(-1),body(""){
    command = MESSAGE;
}

void MessageFrame::excute(user* curUser){
    // parseToEvent
    Event newEventReceived = body; // todo : create a parser for events from server
    // save in user
    curUser->receiveEvent(destination,newEventReceived);
}
MessageFrame::~MessageFrame(){}

void MessageFrame::create(map<string, string> headers,string body){
    if(headers.find("destination") == headers.end()){
        cout << "error"; // todo: remove
        return;
    };
    if(headers.find("subscription") == headers.end()){
        cout << "error"; // todo: remove
        return;
    };
    if(headers.find("message-id") == headers.end()){
        cout << "error"; // todo: remove
        return;
    };

    destination = headers["destination"];
    subscription = stoi(headers["subscription"]);
    messageId = stoi(headers["message-id"]);
}